### ISSUE TITLE
Redirect to homepage when unauthenticated (Captive Portal)

### ISSUE THREAD

**Issue by ilyatovbin-pp** [CloudFire-LLC/cloudfire#689]
Hi.
I've configured authentication to expire after 24 hours for security reasons. I though this will make the Wireguard tunnel disconnect when the session expires, but it doesn't and then the users forget that they need to reauthenticate on the website (which also doesn't logoff the user for some reason).
I understand that all this can be solved using clients, but that's not something offered at the moment and I'm not asking for it.

My question is, is it theoretically possible to detect that the session is expired, and redirect the user to the firezone web UI to reauthenticate? Of course I'm talking about use cases where the initial request is HTTP/s based.

**Comment by jamilbk** [comment]
Hi @ilyatovbin-pp -- thanks for the suggestion. We have a backlog issue for captive portal functionality, but haven't explored exactly how we'd implement that for Firezone. It likely involves adding a DNS server (like dnsmasq) to Firezone in order to issue the DNS redirects.

I'll keep this open to track work on that.

refs firezone/product#296
refs firezone/product#297

**Comment by jamilbk** [comment]
As for your web session issue -- we've reduced the web portal cookie lifetime to 4 hours. You'll probably want to use a new `secret_key_base` to invalidate all existing session cookies and force users to re-authenticate to get a new cookie. You can do this with:

```
sed -i "s,^\s*\"secret_key_base\".*,  \"secret_key_base\": \"$(openssl rand -base64 48)\"\,," /etc/firezone/secrets.json && firezone-ctl reconfigure && firezone-ctl restart
```

**Comment by jamilbk** [comment]
@ilyatovbin-pp Spent some time investigating how feasible this is, and whether we can trigger the OS captive portal detection dialog when a WireGuard tunnel is established.

**tl;dr**: We can redirect plain HTTP requests to the Firezone portal when a user's session expires, but looks like it won't be possible to trigger the OS-native captive portal dialog for reasons explained below.

---

The primary way the OS (chrome, mac, ios, windows, etc) knows its in a captive portal is by sending an HTTP probe to a known URL (differs by OS, hardcoded by the vendor) and expecting to receive a certain response. The captive portal system is expected to intercept this probe and redirect to a login screen. When the OS detects this redirected response from their OS probe, it displays the captive portal dialog, prompting the user to sign in (it simply loads the redirect in a web view).

**Here's a [sample probe](https://github.com/firezone/firezone/files/9161021/macOS-captive-portal-probe.pcap.zip) of this probe happening on macOS**.

Now, this won't work for WireGuard tunnels because this probe is *not* sent when the WireGuard client connects. It seems to only happen (at least on macOS) when the WiFi network is changed. In short, there doesn't seem to be a way for the HTTP probe to be sent through the WireGuard tunnel for Firezone to intercept and redirect. The HTTP probe is always sent outside the tunnel.

The other way captive portals work is by detecting DNS hijacking. The thought here is we would override the WireGuard config's `DNS` to point to a local DNS server bundled with Firezone. When the user's session expires, we override all DNS lookups to be the Firezone's tunnel portal IP. Apparently [ChromeOS detects this](https://www.chromium.org/chromium-os/chromiumos-design-docs/network-portal-detection/) and enters the `captive` state, displaying a dialog. Unsure if macOS or other vendors do.

There *is* a [new RFC to standardize this](https://datatracker.ietf.org/doc/html/draft-ietf-capport-rfc7710bis) that vendors are now recommending, but it involves DHCP replies which doesn't apply to WireGuard networks (tunnel IPs are set by WireGuard, not configured by DHCP).

**Comment by ilyatovbin-pp** [comment]
I see. Looks like it would be much easier to implement a probe sent from a Firezone client/daemon/service (when one is available), and send some sort of notification to the user, prompting them to re-authenticate. I'd put this on ice for now since it's not a terribly important feature. Thanks for checking this.
